src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+
+    struct fd *filedes = &fd_tbl[FD];
+    if (!filedes->file) return OS_FS_ERR_INVALID_FD;
+    if (filedes->ino != filedes->file->ino) return OS_FS_ERR_INVALID_FD;
+    return OS_FS_SUCCESS;
 }
 
-/*
- * parses a loaded tar into fs[0]
- */
-uint32 tar_parse()
+//finds the next free file
+uint32 newfile_get(struct fsobj **o)
 {
-    assert(start && end);
-    assert(size < INT32_MAX);
-    assert(end - start > 0);
-    assert(size == (size_t) (end - start));
-    assert(fs[0].root->filedes == 1); //first fs should be rooted by first file
+    uint32 count = 0;
+    while (count < MAX_NUM_FILES && files[count].ino) {
+        count++;
+    }
+    if (count == MAX_NUM_FILES) return OS_FS_ERROR;
+    *o =  &files[count];
 
-    if (!fs[0].devname) return OS_FS_ERR_DRIVE_NOT_CREATED;
+    **o = (struct fsobj) {
+        //ino needs to be unique and nonzero, so ino is defined as index+1
+        .ino = count +1
+    };
+    return OS_FS_SUCCESS;
+}
 
-    struct fsobj *o = fs[0].root;
-    uint32 offset = 0;
+uint32 file_insert(struct fsobj *o, char *path)
+{
+    //paths should always begin with '/' but we do not need it here
+    if (path[0] != '/') return OS_FS_ERR_PATH_INVALID;
+    path++;
 
-    while (offset + start < end) {
-        //tar ends after two empty records
-        if ( !(offset + start)[0] && !(offset + start)[TAR_BLOCKSIZE]) return OS_FS_SUCCESS;
+    assert(o && path && openfs);
 
-        if (tar_cphdr(offset, o))     return OS_FS_ERR_DRIVE_NOT_CREATED;
-        if (file_insert(o, &fs[0]))   return OS_FS_ERR_DRIVE_NOT_CREATED;
-        //data is alligned to 512 byte blocks.  a header is 500 bytes, and
-        // the file's data begins exactly after the header
-        // therefor the next header is 500 + o->size rounded up to a mult of 512
-        offset += round_to_blocksize(o->size + 500);
-        if (newfile_get(&o))         return OS_FS_ERR_DRIVE_NOT_CREATED;
+    if (!openfs->root) {
+        if (strcmp(o->name, path) != 0) return OS_FS_ERR_PATH_INVALID;
+        openfs->root = o;
+        return OS_FS_SUCCESS;
     }
 
-    //tar ends before two empty records are found
-    return OS_FS_ERROR;
+    assert(openfs->root->ino);
+    struct fsobj *cur = openfs->root;
+
+    // token is the current directory in path
+    char path_temp[OS_MAX_PATH_LEN * 2];
+    strcpy(path_temp, path);
+    const char delim[2] = "/";
+    char *token = strtok(path_temp, delim);
+
+    if (strcmp(token, cur->name) != 0) return OS_FS_ERR_PATH_INVALID;
+    //loop terminates when it finds a place to put o or determines path is invalid
+
+    while (1) {
+        if (token == NULL) {
+            return OS_FS_ERR_PATH_INVALID;
+        }
+        assert(cur->name);
+
+        //if there is no child, then insert as child
+        if (!cur->child) {
+            // if the next part of the path is not o->name or there is a part after it, bad path
+            token = strtok(NULL, delim);
```

Comment:
I don't understand?  But it doesn't matter because I am getting rid of strtok anyway.  But this is a correct use of strtok